Added prefix and route nesting to AppRoutes

DenuxPlays opened the pull request:
replaced snapshot testing with normal asserts:

It was faster to replace snapshots with basic asserts then to look through multiple snapshot files with no relevant content.
Hope thats ok but I could revert it.

Closes: https://github.com/loco-rs/loco/issues/1116

kaplanelad commented on src/controller/app_routes.rs:
```diff
@@ -132,23 +130,134 @@ impl AppRoutes {
+    /// use loco_rs::controller::AppRoutes;
+    ///
+    /// let app_routes = AppRoutes::with_default_routes()
+    ///     .prefix("api")
+    ///     .nest_prefix("v1");
```
Can you provide a bit complete example? 
```rust

# /api/auth/...
# /api/v1/home...
let app_routes = AppRoutes::with_default_routes()
     .prefix("API")
     .add_route(controllers::auth::routes())
     .nest_prefix("v1")
     .add_route(controllers::home::routes())

```

DenuxPlays replied:
No

With your current CI this would fail.
As controllers does not exist in the scope.

kaplanelad replied:
So let's make it work. 
Do you want to share the error?

DenuxPlays replied:
I'll take a break in a few minutes.
I can update the code and commit it then.

DenuxPlays replied:
Sorry didn't had enough time in my break.
We had to tcpdump some shit very cool thing tbh ^^

But once the ci failed you can take a look.
I think it tries to validates the rust code blocks and it will complain that there is no `controller::` available.

kaplanelad commented on src/controller/app_routes.rs:
````diff
@@ -132,23 +130,134 @@ impl AppRoutes {
+    ///
+    /// In the following example, you are adding `api` as a prefix and then nesting a route within it:
+    ///
+    /// ```rust, no_run
````
can you remove the `no_run`?

kaplanelad commented on src/controller/app_routes.rs:
````diff
@@ -132,23 +130,134 @@ impl AppRoutes {
+    ///
+    /// In the following example, you are adding `api` as a prefix and then nesting multiple routes within it:
+    ///
+    /// ```rust, no_run
````
can you remove the `no_run`?

kaplanelad commented on src/controller/app_routes.rs:
```diff
@@ -253,11 +366,23 @@ mod tests {
+            ("/multiple3", vec![Method::PATCH]),
+        ];
+
+        assert_eq!(routes.len(), expected_routes.len());
+        for (i, route) in routes.iter().enumerate() {
+            assert_eq!(route.uri, expected_routes[i.clone()].0);
+            assert_eq!(route.actions, expected_routes[i].1);
```
The test is a bit hard to follow and maintain. It loops over one vector while expecting values from another, making it less intuitive.

Additionally, if the order changes, updating the test becomes cumbersome. I prefer using a snapshot here since it avoids the need to worry about ordering.

DenuxPlays replied:
Okay I'll revert this change.

kaplanelad commented on src/controller/app_routes.rs:
```diff
@@ -206,25 +315,29 @@ impl AppRoutes {
-                format!("{:?} {}", route.actions, route.uri)
-            );
+        let routes = AppRoutes::with_default_routes().collect();
+        let expected_routes = vec![
```
see other comment related to this

DenuxPlays commented:
I think I fixed the doc test.

I've added the `ignore` attribute.
Not sure if this is what we want as it is not exactly recommended but `compile_fail` also isn't a good fit.

If you want you can take a look [here](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html#attributes) to see what attributes are available.

kaplanelad commented:
> I think I fixed the doc test.
> 
> I've added the `ignore` attribute. Not sure if this is what we want as it is not exactly recommended but `compile_fail` isn't a good fit either.
> 
> If you want you can take a look [here](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html#attributes) to see what attributes are available.

No, this isn’t what we want—this was actually my original comment 🙂.
Let me take a look and push a fix to your PR.

kaplanelad commented:
The ignore feature has been removed: https://github.com/loco-rs/loco/pull/1241/commits/b4aa3f6c8b8bb73b7db44914d486d7d31eaf92ca.

Also, I noticed that we don’t have documentation on AppRoutes. Could you please add it here?: https://loco.rs/docs/the-app/controller/

DenuxPlays commented:
Added some docs
